core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractStaticDataNode.java

```diff
@@ -0,0 +1,159 @@
+package ai.timefold.solver.core.impl.bavet.common;
+
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+
+import ai.timefold.solver.core.impl.bavet.NodeNetwork;
+import ai.timefold.solver.core.impl.bavet.common.tuple.AbstractTuple;
+import ai.timefold.solver.core.impl.bavet.common.tuple.RecordingTupleLifecycle;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleLifecycle;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleState;
+import ai.timefold.solver.core.impl.bavet.uni.AbstractForEachUniNode;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public abstract class AbstractStaticDataNode<Tuple_ extends AbstractTuple> extends AbstractNode
+        implements TupleSourceRoot<Object> {
+    private final StaticPropagationQueue<Tuple_> propagationQueue;
+    private final Map<Object, List<Tuple_>> tupleMap = new IdentityHashMap<>(1000);
+    private final NodeNetwork nodeNetwork;
```

Comment:
I'd find a better name here, something like `nestedNodeNetwork`, `independentNodeNetwork`... To signify that this differs from the main node network.